File: script/make/protocol/lua/SkillProtocol.lua
SkillProtocol = {}

function SkillProtocol.encode(offset, protocol, data)
    if protocol == 11701 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 11702 then
        local offset = offset
        local table = {}
        -- 技能ID
        table[offset] = string.pack(">I4", data["skillId"])
        offset = offset + 1
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function SkillProtocol.decode(offset, protocol, data)
    if protocol == 11701 then
        local offset = offset
        -- 技能列表
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- 技能ID
            local skillId = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 技能等级
            local level = string.unpack(">I2", data, offset)
            offset = offset + 2
            list[listIndex] = {skillId = skillId, level = level}
        end
        return {list = list}
    elseif protocol == 11702 then
        local offset = offset
        -- 结果
        local result = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(result)
        return {result = result}
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

File: script/make/protocol/js/SkillProtocol.js
export default class SkillProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 11701: {

                return new DataView(view.buffer.slice(0, offset));
            }
            case 11702: {
                // extend
                while (view.byteLength < offset + 4) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 技能ID
                view.setUint32(offset, data, false);
                offset = offset + 4;
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 11701: {
                // 技能列表
                const data = [];
                let dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--dataLength >= 0) {
                    // 
                    // 技能ID
                    const dataDataSkillId = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 技能等级
                    const dataDataLevel = view.getUint16(offset, false);
                    offset = offset + 2;
                    // object
                    const dataData = {"skillId": dataDataSkillId, "level": dataDataLevel};
                    // add
                    data.push(dataData);
                }
                return data;
            }
            case 11702: {
                // 结果
                const dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                const dataArray = new Uint8Array(view.buffer.slice(offset, offset + dataLength));
                const data = textDecoder.decode(dataArray);
                offset = offset + dataLength;
                return data;
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}